src/app/qgshandlebadlayers.cpp

```diff
@@ -509,3 +479,162 @@ int QgsHandleBadLayers::layerCount()
+  {
+    int idx = mLayerList->item( i, 0 )->data( Qt::UserRole ).toInt();
+    QDomNode &node = const_cast<QDomNode &>( mLayers[ idx ] );
+
+    QTableWidgetItem *item = mLayerList->item( i, 4 );
+    QString datasource = item->text();
+    QString fileName;
+    const QString name { mLayerList->item( i, 0 )->text() };
+    const QString basepath = datasource.left( datasource.lastIndexOf( '/' ) );
+    const QString longName = datasource.mid( datasource.lastIndexOf( '/' ) );
+    if ( longName.lastIndexOf( '|' ) != -1 )
+      fileName = longName.left( longName.lastIndexOf( '|' ) - 1 );
+    else
+      fileName = longName;
+    datasource = checkBasepath( name, basepath, fileName ).replace( fileName, longName );
+
+    bool dataSourceChanged { false };
+    const QString layerId { node.namedItem( QStringLiteral( "id" ) ).toElement().text() };
+    const QString provider { node.namedItem( QStringLiteral( "provider" ) ).toElement().text() };
+
+
+    // Try first to change the datasource of the existing layers, this will
+    // maintain the current status (checked/unchecked) and group
+    if ( QgsProject::instance()->mapLayer( layerId ) )
+    {
+      QgsDataProvider::ProviderOptions options;
+      QgsMapLayer *mapLayer = QgsProject::instance()->mapLayer( layerId );
+      if ( mapLayer )
+      {
+        mapLayer->setDataSource( datasource, name, provider, options );
+        dataSourceChanged = mapLayer->isValid();
+      }
+    }
+
+    if ( !( dataSourceChanged ) )
+    {
+      datasource = findFile( fileName, basepath ).replace( fileName, longName );
+      if ( QgsProject::instance()->mapLayer( layerId ) && !( datasource.isEmpty() ) )
+      {
+        QgsDataProvider::ProviderOptions options;
+        QgsMapLayer *mapLayer = QgsProject::instance()->mapLayer( layerId );
+        if ( mapLayer )
+        {
+          mapLayer->setDataSource( datasource, name, provider, options );
+          dataSourceChanged = mapLayer->isValid();
+        }
+      }
+      if ( dataSourceChanged )
+      {
+        const QString altBasepath = datasource.left( datasource.lastIndexOf( '/' ) );
+        checkBasepath( name, altBasepath, fileName ).replace( fileName, longName );
+      }
+    }
+
+    // If the data source was changed successfully, remove the bad layer from the dialog
+    // otherwise, try to set the new datasource in the XML node and reload the layer,
+    // finally marks with red all remaining bad layers.
+    if ( dataSourceChanged )
+    {
+      item->setForeground( QBrush( Qt::green ) );
+    }
+    else
+    {
+      node.namedItem( QStringLiteral( "datasource" ) ).toElement().firstChild().toText().setData( datasource );
+      if ( QgsProject::instance()->readLayer( node ) )
+      {
+        mLayerList->removeRow( i-- );
+      }
+      else
+      {
+        item->setForeground( QBrush( Qt::red ) );
+      }
+    }
+  }
+
+  // Final cleanup: remove any bad layer (it should not be any btw)
```

Comment:
Is this something that needs to be done in auto find?
I would imagine the workflow to be:

* click auto find
* hope it fixes things
* fix any leftovers manually

Is it me missing something or are all remaining layers (e.g. network based ones) removed as soon as auto find is clicked?

Reply:
I used the existing logic in apply() as a base. I assume that this could be removed as the cleanup will occur in apply or accept.